python/pdo/contract/replication.py

```diff
@@ -0,0 +1,214 @@
+        
+        try:
+            response = client.store_blocks(block_data_list, expiration)
+            if response is None :
+                Replicator.__queue_of_SSURLS_raising_exceptions__.put(client.ServiceURL) 
+                logger.warn("Adding Storage Service at %s to the set of  bad ones : block_store_list is None", str(client.ServiceURL))
+        except Exception as e:
+            Replicator.__queue_of_SSURLS_raising_exceptions__.put(client.ServiceURL)
+            logger.warn("Adding Storage Service at %s to the set of  bad ones : uknown exception : %s ", str(client.ServiceURL), str(e))
+            response = None
+
+        return response
+    
+    # -----------------------------------------------------------------
+
+    def replicate_change_set(self):
+        """ Replicate change set to the set of reliable storage services."""
+            
+        # identify if replication can be skipped, including the case where there is only one provisioned enclave
+        if len(self.storage_clients_to_replicate) == 0:
+            logger.info('Skipping replication: Only one provisioned enclave, so nothing to replicate')
+            return
+       
+        if len(self.blocks_to_replicate) == 0:
+            logger.info('Skipping replication: No change set, so nothing to replicate')
+            return
+
+        # submit individual replication tasks to various storage services asynchronously. Each replica is handled by a separate thread 
+        executor = concurrent.futures.ThreadPoolExecutor(max_workers = len(self.storage_clients_to_replicate))
+        futures_to_replication_tasks = dict() # to keep track of the tasks
+        expiration = self.availability_duration
+        for index, client in enumerate(self.storage_clients_to_replicate):
+            # get data generator. Optimization: If  this generator can be wrapped to be threadsafe, itertools.tee can be used to create
+            # copies without having to read the cache multiple times. Directly using itertools.tee will not work with multi-threadeding
+            block_data_list = ContractState.block_data_generator(self.contract_id, self.blocks_to_replicate, self.data_dir)
+            future = executor.submit(self.push_blocks_to_storage_service, client, block_data_list, expiration)
+            futures_to_replication_tasks[future] = client.ServiceURL
+       
+        #poll for completed replication tasks, wait until num_provable_replicas number of tasks finish
+        urls_of_successful_storage_services = [self.sservice_url_for_contract] #successful by default
+        num_successful_replicas = 1
+        num_unsuccessful_replicas = 0
+        max_allowable_num_unsucessful_replicas = len(self.storage_clients_to_replicate) - self.num_provable_replicas + 1
+            # Add 1 since storage_clients_to_replicate does not contain the sservice corresponding to the contract enclave
```

Comment:
I guess as soon as you add retrieving the proof (and telling service on hold-time-expections) you will though also have to deal with the that service?  But then i guess that might be best dealt with with a second queue with services you pushed the data but haven't received the receipt yet?  Anyway, something for another PR ...